### PR TITLE
[stdlib] Remove workaround for sort optimization

### PR DESCRIPTION
**Pull request by HassanElDesouky**
Remove workaround for sort optimization. Cherry picked from #40056.

Resolves [SR-14750](https://bugs.swift.org/browse/SR-14750).

cc @lorentey, @glessard, @natecook1000

### CODE REVIEW

File: stdlib/public/core/Sort.swift [hunk 1]
```diff
@@ -344,7 +344,7 @@ internal func _merge<Element>(
   high: UnsafeMutablePointer<Element>,
   buffer: UnsafeMutablePointer<Element>,
   by areInIncreasingOrder: (Element, Element) throws -> Bool
-) rethrows -> Bool {
```
Comment (LucianoPAlmeida) [inline stdlib/public/core/Sort.swift:347]: Although this is internal, since its `inlinable` this is part of stdlib ABI so not sure this can be changed

**Comment by glessard** [comment]
@swift-ci please test macOS platform

**Comment by HassanElDesouky** [comment]
@glessard can you re-run the tests again. There was a merge conflict that I just resolved.

**Comment by glessard** [comment]
@swift-ci please test macOS platform

**Comment by glessard** [comment]
@swift-ci please test Linux platform

**Comment by glessard** [comment]
@natecook1000 @lorentey I'm not sure what to make of the passed tests: it seems like this should register as an ABI change. Thoughts?

**Comment by glessard** [comment]
If we don't merge this, an alternative is to at least add the JIRA bug number: https://github.com/apple/swift/pull/40138

**Comment by glessard** [comment]
@HassanElDesouky The addition of the SR bug number was merged to main, and we have a merge conflict again. It's preferable to solve the conflict by rebasing rather than merging (it makes for a more understandable history.) Thanks!

**Comment by HassanElDesouky** [comment]
@glessard Thank you for adding the SR bug number. I just rebased the branch and fixed the merge conflict.

**Comment by glessard** [comment]
@swift-ci Please Build Toolchain Linux Platform

**Comment by HassanElDesouky** [comment]
@glessard any update whether we're going to land this or not?

**Comment by lorentey** [comment]
Luciano's note on ABI stability isn't resolved yet! Sadly I believe we'll need to preserve the Bool return value as these functions are ABI. (Otherwise we would introduce binary compatibility issues.)

**Comment by glessard** [comment]
We could perhaps remove the workarounds without changing the ABI: simply return true, and add the `@discardableResult` attribute. If nothing else, it would be a minor win in code size.

**Comment by HassanElDesouky** [comment]
@glessard Nice idea! I'm just traveling now so I'll do this solution once I get back.

File: stdlib/public/core/Sort.swift [hunk 2]
```diff
@@ -337,6 +337,7 @@ extension MutableCollection where Self: BidirectionalCollection {
 /// - Precondition: `buffer` must point to a region of memory at least as large
 ///   as `min(mid - lo, hi - mid)`.
 /// - Postcondition: `lo..<hi` is sorted according to `areInIncreasingOrder`.
+@discardableResult
```
Comment (HassanElDesouky) [inline stdlib/public/core/Sort.swift:340]: @glessard I'm sorry, I forgot to ping you to review. I added `@discardableResult` to `_merge` as @LucianoPAlmeida suggested.

Comment (HassanElDesouky) [inline stdlib/public/core/Sort.swift:340]: @glessard I'm sorry, I forget to ping you. I added `@discardableResult` to `_merge` as @LucianoPAlmeida pointed out.

File: stdlib/public/core/Sort.swift [hunk 3]
```diff
@@ -518,13 +518,13 @@ extension UnsafeMutableBufferPointer {
     at i: Int,
     buffer: UnsafeMutablePointer<Element>,
     by areInIncreasingOrder: (Element, Element) throws -> Bool
-  ) rethrows -> Bool {
```
Comment (LucianoPAlmeida) [inline stdlib/public/core/Sort.swift:521]: As far as I understand all ABI-public(`public`, `open`, `@usableFromInline`, or `@inlinable`) functions should not change signature to preserve ABI. In my limited understanding we are allowed to change the implementation details bu not the API.

Comment (HassanElDesouky) [reply stdlib/public/core/Sort.swift:521]: Ahaa. That actually makes sense, thank you @LucianoPAlmeida as always ❤️

File: stdlib/public/core/Sort.swift (hunk 2 above)
Comment (glessard) [reply stdlib/public/core/Sort.swift:340]: No problem, I was in no condition to contribute until today.

File: stdlib/public/core/Sort.swift [hunk 4]
```diff
@@ -550,7 +547,7 @@ extension UnsafeMutableBufferPointer {
     _ runs: inout [Range<Index>],
     buffer: UnsafeMutablePointer<Element>,
     by areInIncreasingOrder: (Element, Element) throws -> Bool
-  ) rethrows -> Bool {
+  ) rethrows {
```
Comment (glessard) [inline stdlib/public/core/Sort.swift:550]: Must return Bool, same as above. Can have @discardableResult.

File: stdlib/public/core/Sort.swift [hunk 5]
```diff
@@ -625,14 +617,11 @@ extension UnsafeMutableBufferPointer {
     _ runs: inout [Range<Index>],
     buffer: UnsafeMutablePointer<Element>,
     by areInIncreasingOrder: (Element, Element) throws -> Bool
-  ) rethrows -> Bool {
-    // FIXME: Remove this, it works around SR-14750 (rdar://45044610)
-    var result = true
+  ) rethrows {
```
Comment (glessard) [inline stdlib/public/core/Sort.swift:620]: Must return Bool, same as above. Can have @discardableResult.

**Review by glessard — CHANGES_REQUESTED** [review]

**Review by glessard — COMMENTED** [review]
Looks good. I'll trigger a tests run.

**Comment by glessard** [comment]
@swift-ci please test

**Review by glessard — APPROVED** [review]

**Comment by xwu** [comment]
@swift-ci please benchmark

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test Windows Platform

**Comment by glessard** [comment]
The test failure was in a test related to the move-only work. Seems to fail for a few different PRs.

**Comment by glessard** [comment]
https://github.com/apple/swift/pull/40806 should fix the test failure above.

**Comment by glessard** [comment]
@swift-ci please test macOS platform

**Comment by xwu** [comment]
@swift-ci please test macOS platform

**Comment by HassanElDesouky** [comment]
Tests passed 🎉
@glessard

**Comment by glessard** [comment]
Thanks @HassanElDesouky for working on this!

**Pull request merged**